### DISCUSSION TITLE
Reproducing PSGTR and PSGFormer

### DISCUSSION THREAD

**Discussion by Jingkang50** [Jingkang50/OpenPSG#51] [General]
If we want to reproduce the results of the paper, we need to change https://github.com/Jingkang50/OpenPSG/blob/a0e5c8f92b7c27883c545ae5f684584a5fa8b351/openpsg/models/frameworks/psgtr.py#L16 to eval_pan_rels=False.
The difference between the evaluation of paper and the competition is that:
in the paper, the mask of each triplet can be generated independently, but the competition needs to submit a panseg png file, so that the mask of the triplet comes from a unified panseg, so we set `eval_pan_rels=True`.
We have provided the log file in this issue https://github.com/Jingkang50/OpenPSG/issues/33 for your reference. 
We are using the release psgtr code to run the results of the psgtr in the paper. We have compared the logs that other students have not reproduced before, probably because the segmentation loss of psgtr has not converged, it may be recommended to add the focal loss of seg, although we did not have it to achieve the results of the paper.
The r50 and 100 of psgtr are much higher than the corresponding values ​​of R@20. 
We think that in the case of `eval_pan_rels=False`, psgtr has a higher degree of freedom and is not limited by the effect of the panoptic segmentation.
For example, the advantage of psgtr is that a mask can be tagged by many categories in different triplets, which makes it easier to guess correctly. But this advantage is weakened in the competition based on a unified panseg map after fusion.

**Comment by ShunchiZhang** [comment]
I'm still confused about it...

It seems that both

1. changing the following line to `eval_pan_rels=False`

https://github.com/Jingkang50/OpenPSG/blob/a0e5c8f92b7c27883c545ae5f684584a5fa8b351/openpsg/models/frameworks/psgtr.py#L16

2. deleting `detection_method='pan_seg'` in the configuration file.

only affects the evaluation process rather than the training process.

I cannot see how these explain the difficulty of reproducing the reported results.

**Reply by ShunchiZhang** [reply]
Also, the reported results are based on the IoU of panoptic masks instead of bounding boxes, which is in accord with the competition. The only reason I can guess is that the training dataset for the competition is smaller.

**Reply by Jingkang50** [reply]
Yes, `eval_pan_rels` and  `detection_method` only affect the evaluation process, but different evaluation protocol will bring different scores.
To be more specific, `eval_pan_rels=False` means every triplet can independently generate subject and object masks, while `eval_pan_rels=True` requires every triplet to **select** masks from the fused, unified panoptic segmentation mask. It will downgrade the performance of one-stage models PSGTR and PSGFormer compared to the paper results.
For the sake of the submission file size, the competition requires a unified panseg for each image, rather than two masks per triplet, so the current default `eval_pan_rels` is set `True`, which is different from the paper's setting.
Also, the test set is not fully released, which will also affect the results. The full test set will be released in the mid Oct.

All the training data is provided.

**Reply by ShunchiZhang** [reply]
Thanks for this really detailed explanation! I will change the setting and evaluate again.

**Reply by ShunchiZhang** [reply]
Hi @Jingkang50, after changing `eval_pan_rels` into `False`, I successfully reproduce the results on 8*V100! Thank you guys so much for your excellent work and patient and detailed explanations!

**Reply by Jingkang50** [reply]
Glad to hear that! Thank you so much for your great questions, which will for sure help many other followers!

**Comment by ShunchiZhang** [comment]
Another question: the segmentation loss described in the paper is `DICE/F-1 Loss`, while the implementation seems to show that the segmentation loss is the combination of `L1Loss`, `GIoULoss`, and `psgtrDiceLoss`. Will this discrepancy make any difference?

https://github.com/Jingkang50/OpenPSG/blob/a0e5c8f92b7c27883c545ae5f684584a5fa8b351/configs/_base_/models/psgtr_r50.py#L52-L67

https://github.com/Jingkang50/OpenPSG/blob/a0e5c8f92b7c27883c545ae5f684584a5fa8b351/openpsg/models/relation_heads/psgtr_head.py#L410-L440

**Reply by GSeanCDAT** [reply]
Hi Shunchi,
Thanks for your continuous interest in our work!
1. In our implementation of PSGTR, we predict bounding boxes as well as segmentation masks, and the `L1Loss` and `GIoULoss` are losses for bounding boxes. Those losses can help optimize our output triplet queries and thus help the segmentation prediction.
2. Besides helping segmentation indirectly, those bounding boxes prediction are used in our Hungarian Triplet Matcher. Compared with using segmentation masks in the matching operation which are not computably affordable, those bounding-box triplets can literally reduce computations.
3. As for `psgtrDiceLoss`, this is just our implementation of `DICE Loss`.
Our model is trained with those three losses and we get the results reported in our paper from the model. Thus there are neither discrepancies nor differences in our practice.

**Reply by ShunchiZhang** [reply]
Thanks for the comprehensive reply! I am much clearer now.

**Comment by sirosann** [comment]
I tried to turn the `eval_pan_rels` to Flase,But I found this to be slow to val after training epoch, and the process would automatically be killed after a few epochs(often in 1th or 3th epoch).
The following is the log after being stopped from training.By the way I trained with 8*NVIDIA RTX A6000. 
<img width="1802" alt="屏幕截图 2023-04-16 194255" src="https://user-images.githubusercontent.com/115337366/232438851-25476a12-5bcb-41d1-9ece-4762aa577fe7.png">
